### PR TITLE
kube-apiserver: default --enable-logs-handler flag to false

### PR DESCRIPTION
**Pull request by pandaamanda**
#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:

ref #77611

#### Does this PR introduce a user-facing change?

```release-note
ACTION REQUIRED: kube-apiserver: `--enable-logs-handler` now defaults to false, which disables the `/logs` endpoint served by kube-apiserver. If serving log files within the `/var/log` filepath at the `/logs` endpoint is still desired, the functionality is now opt-in and must be enabled by explicitly setting `--enable-logs-handler=true`
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by Jefftree** [comment]
Other than the /logs change, are the other OpenAPI changes related to this PR?

**Comment by pandaamanda** [comment]
> Other than the /logs change, are the other OpenAPI changes related to this PR?

only /logs api. the other openapi changes generate when i exec `ack/update-openapi-spec.sh` since the pull-kubernetes-verify test failed and says `/home/prow/go/src/k8s.io/kubernetes/api/openapi-spec is out of date. Please run hack/update-openapi-spec.sh}`. but it seems not work.

**Comment by Jefftree** [comment]
> > Other than the /logs change, are the other OpenAPI changes related to this PR?
> 
> only /logs api. the other openapi changes generate when i exec `ack/update-openapi-spec.sh` since the pull-kubernetes-verify test failed and says `/home/prow/go/src/k8s.io/kubernetes/api/openapi-spec is out of date. Please run hack/update-openapi-spec.sh}`. but it seems not work.

Interesting, can you rebase off the latest master branch and then run update-openapi-spec?

**Comment by pandaamanda** [comment]
> > > Other than the /logs change, are the other OpenAPI changes related to this PR?
> > 
> > 
> > only /logs api. the other openapi changes generate when i exec `ack/update-openapi-spec.sh` since the pull-kubernetes-verify test failed and says `/home/prow/go/src/k8s.io/kubernetes/api/openapi-spec is out of date. Please run hack/update-openapi-spec.sh}`. but it seems not work.
> 
> Interesting, can you rebase off the latest master branch and then run update-openapi-spec?

when i checkout to master branch and is up to date with 'origin/master', i exec `./hack/update-openapi-spec.sh` and the result is same.

```
root@xzl-dev:/home/code/src/k8s.io/kubernetes# git checkout master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
root@xzl-dev:/home/code/src/k8s.io/kubernetes# git pull -r upstream master
From https://github.com/kubernetes/kubernetes
 * branch                    master     -> FETCH_HEAD
Already up to date.
Current branch master is up to date.
root@xzl-dev:/home/code/src/k8s.io/kubernetes# ./hack/update-openapi-spec.sh
make: Entering directory '/home/code/src/k8s.io/kubernetes'
make[1]: Entering directory '/home/code/src/k8s.io/kubernetes'
...
...
+++ [0624 02:55:36] SUCCESS
+++ [0624 02:55:37] Clean up complete
root@xzl-dev:/home/code/src/k8s.io/kubernetes# git status
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   api/openapi-spec/swagger.json

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        api/openapi-spec/v3/apis__autoscaling__v2beta1_openapi.json
        api/openapi-spec/v3/apis__batch__v1beta1_openapi.json
        api/openapi-spec/v3/apis__discovery.k8s.io__v1beta1_openapi.json
        api/openapi-spec/v3/apis__events.k8s.io__v1beta1_openapi.json
        api/openapi-spec/v3/apis__flowcontrol.apiserver.k8s.io__v1alpha1_openapi.json
        api/openapi-spec/v3/apis__node.k8s.io__v1beta1_openapi.json
        api/openapi-spec/v3/apis__policy__v1beta1_openapi.json
        api/openapi-spec/v3/apis__storage.k8s.io__v1alpha1_openapi.json

no changes added to commit (use "git add" and/or "git commit -a")
```

File: test/e2e/network/networking.go [hunk 1]
```diff
@@ -115,7 +115,7 @@ var _ = common.SIGDescribe("Networking", func() {
 			{path: "/version"},
 			// TODO: test proxy links here
 		}
-		if !framework.ProviderIs("gke", "skeleton") {
+		if !framework.ProviderIs("gce", "gke", "skeleton") {
```
Comment (Jefftree) [inline test/e2e/network/networking.go:118]: Is this change part of the PR?

**Review by Jefftree — COMMENTED** [review]
Diff looks much better, one comment otherwise lgtm

File: test/e2e/network/networking.go (hunk 1 above)
Comment (pandaamanda) [reply test/e2e/network/networking.go]: fixed.

**Comment by liggitt** [comment]
I don't think we have a clear understanding of who is depending on this functionality today, and whether they are aware of the announced deprecation and scheduled removal. Since the flag defaults to `true`, I suspect that *many* people are relying on this and have no idea it was marked as deprecated. Since it defaults to `true`, we also can't get a sense of how much usage this has based on [looking for people enabling the CLI flag](https://grep.app/search?q=--enable-logs-handler).

If we want to improve the security posture of `kube-apiserver`, defaulting this to `false` with a prominent release note seems like a good intermediate step. That gives people who are relying on the functionality notice at the 1.25 release boundary, but still gives them a way to access the functionality if they rely on it. People explicitly opting into the option would also make usage much more visible, so we could determine if the disruption caused by the removal is too high.

My inclination is to turn it off by default but leave the capability in place as opt-in for compatibility.

**Comment by Jefftree** [comment]
That makes sense. Defaulting to off will bring more awareness to the deprecation but still allow users to opt in if this is a feature they rely on. 

@pandaamanda Would you like to make this change? We'll keep the logs endpoint and API but change the [default](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/options/options.go#L112) to false and maybe add a note somewhere indicating that this will be removed in a future release. (Does 1.28 sound reasonable @liggitt?)

**Comment by liggitt** [comment]
> We'll keep the logs endpoint and API but change the [default](https://github.com/kubernetes/kubernetes/blob/master/cmd/kube-apiserver/app/options/options.go#L112) to false and maybe add a note somewhere indicating that this will be removed in a future release. (Does 1.28 sound reasonable @liggitt?)

For something like this that has been treated as stable in every version without a clear version / compatibility guarantee, I think we need to be informed by use before setting a target removal release (or even being sure we can/should remove it). The context for that should be broader than a single PR (e.g. tagging in api-server maintainers, sig-security folks, and figuring out how to assess use and impact of removal, etc)

**Comment by liggitt** [comment]
Following the chain back to the [originating comment](https://github.com/kubernetes/kubernetes/issues/77182#issuecomment-487439173), what I was proposed was changing the default to false, not removing the functionality entirely.

**Comment by pandaamanda** [comment]
> Following the chain back to the [originating comment](https://github.com/kubernetes/kubernetes/issues/77182#issuecomment-487439173), what I was proposed was changing the default to false, not removing the functionality entirely.

Should we make an announcement before change the default value? Thks.

**Comment by liggitt** [comment]
If all we are doing is changing the default for the CLI flag but are retaining the functionality, from my perspective, highlighting the change in the release notes, making any relevant docs updates, and sending an announcement to the dev list in parallel with merging this change is sufficient.

sig-security might have other places to recommend highlighting this as a configuration option with security implications.
/sig security

File: cmd/kube-apiserver/app/options/options.go [hunk 2]
```diff
@@ -168,10 +166,6 @@ func (s *ServerRunOptions) Flags() (fss cliflag.NamedFlagSets) {
 	fs.BoolVar(&s.AllowPrivileged, "allow-privileged", s.AllowPrivileged,
 		"If true, allow privileged containers. [default=false]")
 
-	fs.BoolVar(&s.EnableLogsHandler, "enable-logs-handler", s.EnableLogsHandler,
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go:171]: restore this line so the flag can be set true if desired

File: cmd/kube-apiserver/app/server.go [hunk 3]
```diff
@@ -269,7 +269,6 @@ func CreateKubeAPIServerConfig(s completedServerRunOptions) (
 			StorageFactory:          storageFactory,
 			EventTTL:                s.EventTTL,
 			KubeletClientConfig:     s.KubeletConfig,
-			EnableLogsSupport:       s.EnableLogsHandler,
```
Comment (liggitt) [inline cmd/kube-apiserver/app/server.go:272]: revert this change (and all the other changes that plumb and set up handlers based on this value)

File: cmd/kube-apiserver/app/options/options.go [hunk 4]
```diff
@@ -109,7 +108,6 @@ func NewServerRunOptions() *ServerRunOptions {
 		Logs:                    logs.NewOptions(),
 		Traces:                  genericoptions.NewTracingOptions(),
 
-		EnableLogsHandler:                 true,
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go:112]: I think this PR should essentially reduce to changing this line from true to false

**Comment by liggitt** [comment]
oh, looks like the `hack/update-openapi-spec.sh` and e2e test that expected the /logs API path did still need adjusting with this defaulting off

**Comment by liggitt** [comment]
the test failures are legitimate, the generated openapi spec and e2e test need adjusting

**Comment by pandaamanda** [comment]
> the test failures are legitimate, the generated openapi spec and e2e test need adjusting

It seems a flake. last time it appeared.

**Comment by liggitt** [comment]
This lgtm. I would like an ack from an api-machinery lead on the default change.

/assign @deads2k @lavalamp 

cc @kubernetes/sig-security-pr-reviews for visibility to the change to default to not exposing API server machine logs

cc @kubernetes/sig-cluster-lifecycle @neolit123 for visibility to kubeadm... would you expect kubeadm to go with the new default-off value?

**Comment by lavalamp** [comment]
Sorry I didn't see this until now. I think users need a release of warning before the default changes. It is reasonable to have interpreted the deprecation notice as meaning that the feature would be on all the time when the flag was removed. Additionally I'm not sure what to do about the openapi -- I kind of think we should publish it even if the flag defaults off.

**Comment by sftim** [comment]
If the PR for https://github.com/kubernetes/website/issues/38870 lands first, we should update those docs when this goes in.

**Comment by sftim** [comment]
Relevant to https://github.com/kubernetes/kubernetes/issues/81125

**Comment by liggitt** [comment]
FWIW, there's a KEP in progress to expand access to host logs on the kubelet side, which doesn't seem coherent with the goal of reducing access to host logs via the apiserver... commented at https://github.com/kubernetes/kubernetes/pull/96120#pullrequestreview-1262371094 and cross-linked here.

**Comment by jayunit100** [comment]
this PR doesnt have an open issue assocaited with it.  Can we fix that ? Otherwise, theres no way to evaluate wether theres decision precedence that will blog the node logs were trying to get in #96120
